src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs

```diff
@@ -410,95 +433,180 @@ protected override bool ReceivePluginInternal(object message)
-            }
-            subscriber.Tell(new CurrentPersistenceIds(GetAllPersistenceIds()));
+            var limitValue = replay.Max >= int.MaxValue ? int.MaxValue : (int)replay.Max;
+
+            var fromSequenceNr = replay.FromOffset;
+            var toSequenceNr = replay.ToOffset;
+            var builder = Builders<JournalEntry>.Filter;
+
+            var seqNoFilter = builder.Empty;
+            if (fromSequenceNr > 0)
+                seqNoFilter &= builder.Gt(x => x.Ordering, new BsonTimestamp(fromSequenceNr));
+            if (toSequenceNr != long.MaxValue)
+                seqNoFilter &= builder.Lte(x => x.Ordering, new BsonTimestamp(toSequenceNr));
+
+
+            // Need to know what the highest seqNo of this query will be
+            // and return that as part of the RecoverySuccess message
+            var maxSeqNoEntry = await _journalCollection.Value.Find(seqNoFilter)
+                .SortByDescending(x => x.Ordering)
+                .Limit(1)
+                .SingleOrDefaultAsync();
+
+            if (maxSeqNoEntry == null)
+                return 0L; // recovered nothing
+
+            var maxOrderingId = maxSeqNoEntry.Ordering.Value;
+            var toSeqNo = Math.Min(toSequenceNr, maxOrderingId);
+
+            var readFilter = builder.Empty;
+            if (fromSequenceNr > 0)
+                readFilter &= builder.Gt(x => x.Ordering, new BsonTimestamp(fromSequenceNr));
+            if (toSequenceNr != long.MaxValue)
+                readFilter &= builder.Lte(x => x.Ordering, new BsonTimestamp(toSeqNo));
+            var sort = Builders<JournalEntry>.Sort.Ascending(x => x.Ordering);
+
+            await _journalCollection.Value.Find(readFilter)
+                .Sort(sort)
+                .Limit(limitValue)
+                .ForEachAsync(entry =>
+                {
+                    var persistent = ToPersistenceRepresentation(entry, ActorRefs.NoSender);
+                    foreach (var adapted in AdaptFromJournal(persistent))
+                    {
+                        replay.ReplyTo.Tell(new ReplayedEvent(adapted, entry.Ordering.Value), ActorRefs.NoSender);
+                    }
+                });
+            return maxOrderingId;
         }
 
         private void AddTagSubscriber(IActorRef subscriber, string tag)
         {
             if (!_tagSubscribers.TryGetValue(tag, out var subscriptions))
             {
-                subscriptions = new HashSet<IActorRef>();
-                _tagSubscribers.Add(tag, subscriptions);
+                _tagSubscribers = _tagSubscribers.Add(tag, ImmutableHashSet.Create(subscriber));
+            }
+            else
+            {
+                _tagSubscribers = _tagSubscribers.SetItem(tag, subscriptions.Add(subscriber));
             }
-
-            subscriptions.Add(subscriber);
         }
 
         private IEnumerable<string> GetAllPersistenceIds()
         {
             return _journalCollection.Value.AsQueryable()
```

Comment:
Change this to query the `_metadataCollection` instead - that will be a tremendously less expensive query than scanning the entire journal.